src/plugins/compression.cpp

```diff
@@ -0,0 +1,91 @@
+#include "compression.h"
+
+std::unique_ptr<Compression> Compression::create(std::string encoding) {
+#ifdef ENABLE_LZ4
+    if (encoding.compare("LZ4") == 0) {
+        return std::make_unique<Lz4Compression>();
+    }
+    else
+#endif /* ENABLE_LZ4 */
+    if (encoding.compare("None") == 0) {
+        return std::make_unique<NoCompression>();
+    }
+    else {
+        return std::unique_ptr<Compression>(nullptr);
+    }
+}
+
+#ifdef ENABLE_LZ4
+int Lz4Compression::compress(ByteIterator uncompressedStart, ByteIterator uncompressedEnd, std::vector<char>& compressed) {
+    uncompressedSize = uncompressedEnd-uncompressedStart;
+    compressed.resize(uncompressedSize+sizeof(size_t));
```

Comment:
why do you need this `+sizeof(size_t)` here?

Reply:
I add the uncompressed size to the beginning of the actual data as meta data.
This way, the correct size for a decompression buffer can be allocated without either transferring extra meta data (as e.g. the various file formats do), or even worse, allocating larger memory blocks heuristically until there is no decompression error. The uncompressed size is always in the first size_t bytes.

Reply:
ok, seems reasonable